src/metabase/models/field.clj

```diff
@@ -166,22 +161,38 @@
   see), would require only a few megs of RAM, and again only if every single Table was looked up in a span of 5
   seconds."
   (memoize/ttl
-   ^{::memoize/args-fn (fn [[table-id read-or-write]]
-                         [(mdb.connection/unique-identifier) table-id read-or-write])}
-   (fn [table-id read-or-write]
-     (let [{schema :schema, db-id :db_id} (t2/select-one ['Table :schema :db_id] :id table-id)]
-       (perms-objects-set* db-id schema table-id read-or-write)))
+   ^{::memoize/args-fn (fn [[table-id]]
+                         [(mdb.connection/unique-identifier) table-id])}
+   (fn [table-id]
+     (:db_id (t2/select-one [:model/Table :db_id] :id table-id)))
    :ttl/threshold 5000))
 
-;;; Calculate set of permissions required to access a Field. For the time being permissions to access a Field are the
-;;; same as permissions to access its parent Table.
-(defmethod mi/perms-objects-set :model/Field
-  [{table-id :table_id, {db-id :db_id, schema :schema} :table} read-or-write]
-  (if db-id
-    ;; if Field already has a hydrated `:table`, then just use that to generate perms set (no DB calls required)
-    (perms-objects-set* db-id schema table-id read-or-write)
-    ;; otherwise we need to fetch additional info about Field's Table. This is cached for 5 seconds (see above)
-    (cached-perms-object-set table-id read-or-write)))
+(defn- field->db-id
+  [{table-id :table_id, {db-id :db_id} :table}]
+  (or db-id (cached-db-id table-id)))
+
+(defmethod mi/can-read? :model/Field
+  ([instance]
+   (= :unrestricted
+      (data-perms/table-permission-for-user
+       api/*current-user-id*
+       :perms/data-access
+       (field->db-id instance)
+       (:table_id instance))))
+  ([model pk]
+   (mi/can-read? (t2/select-one model pk))))
+
+(defmethod mi/can-write? :model/Field
+  ([instance]
+   (if (premium-features/enable-advanced-permissions?)
+     (= :yes (data-perms/table-permission-for-user
+              api/*current-user-id*
+              :perms/manage-table-metadata
+              (field->db-id instance)
+              (:table_id instance)))
+     (mi/superuser?)))
```

Comment:
I think this check should be moved to EE code. Currently `perms/data-model-write-perms-path` tries to resolve `metabase-enterprise.advanced-permissions.models.permissions/data-model-write-perms-path` and calls into it when EE functionality is enabled; otherwise it requires admin perms to write fields.

Could you do the same here, maybe with a `defenterprise` macro? We should make sure that enforcement of EE-only permissions is done in EE namespaces so that it's covered by the correct license.